datashader/spatial/proximity.py

```diff
@@ -0,0 +1,374 @@
+
+    for line in prange(height):
+        # Read for target values.
+        for i in prange(width):
+            scan_line[i] = img[line][i]
+
+        line_proximity = np.zeros(width, dtype=np.float64)
+        for i in prange(width):
+            line_proximity[i] = -1.0
+
+        # left to right
+        _process_proximity_line(scan_line, x_coords, y_coords,
+                                pan_near_x, pan_near_y, True, line,
+                                width, max_distance, line_proximity,
+                                target_values, distance_metric)
+
+        # right to left
+        _process_proximity_line(scan_line, x_coords, y_coords,
+                                pan_near_x, pan_near_y, False, line,
+                                width, max_distance, line_proximity,
+                                target_values, distance_metric)
+
+        for i in prange(width):
+            img_proximity[line][i] = line_proximity[i]
+
+    # Loop from bottom to top of the image.
+    for i in prange(width):
+        pan_near_x[i] = -1
+        pan_near_y[i] = -1
+
+    for line in prange(height-1, -1, -1):
+        # Read first pass proximity.
+        for i in prange(width):
+            line_proximity[i] = img_proximity[line][i]
+
+        # Read pixel target_values.
+        for i in prange(width):
+            scan_line[i] = img[line][i]
+
+        # Right to left
+        _process_proximity_line(scan_line, x_coords, y_coords,
+                                pan_near_x, pan_near_y, False, line,
+                                width, max_distance, line_proximity,
+                                target_values, distance_metric)
+
+        # Left to right
+        _process_proximity_line(scan_line, x_coords, y_coords,
+                                pan_near_x, pan_near_y, True, line,
+                                width, max_distance, line_proximity,
+                                target_values, distance_metric)
+
+        # final post processing of distances
+        for i in prange(width):
+            if line_proximity[i] < 0 or np.isnan(scan_line[i]):
+                # this corresponds the the nan value of input raster.
+                line_proximity[i] = np.nan
+
+        for i in prange(width):
+            img_proximity[line][i] = line_proximity[i]
+    return img_proximity
+
+
+# ported from
+# https://github.com/OSGeo/gdal/blob/master/gdal/alg/gdalproximity.cpp
+def proximity(raster, target_values=[], distance_metric='EUCLIDEAN'):
+    """Compute the proximity of all pixels in the image to a set of pixels in
+    the source image.
+
+    This function attempts to compute the proximity of all pixels in the
+    image to a set of pixels in the source image. The following options are
+    used to define the behavior of the function. By default all non-zero pixels
+    in ``raster.values`` will be considered the "target", and all proximities
+    will be computed in pixels.  Note that target pixels are set to the value
```

Comment:
Don't we have the actual distance units from the xarray metadata? Shouldn't we report distances in those units?